node/core/candidate-selection/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+	//
+	// this function is in charge of creating and executing the job's main loop
+	fn run(
+		_relay_parent: Hash,
+		_run_args: Self::RunArgs,
+		metrics: Self::Metrics,
+		receiver: mpsc::Receiver<ToJob>,
+		sender: mpsc::Sender<FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
+		async move {
+			let job = CandidateSelectionJob::new(metrics, sender, receiver);
+
+			// it isn't necessary to break run_loop into its own function,
+			// but it's convenient to separate the concerns in this way
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl CandidateSelectionJob {
+	pub fn new(
+		metrics: Metrics,
+		sender: mpsc::Sender<FromJob>,
+		receiver: mpsc::Receiver<ToJob>,
+	) -> Self {
+		Self {
+			sender,
+			receiver,
+			metrics,
+			seconded_candidate: None,
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			match msg {
+				ToJob::CandidateSelection(CandidateSelectionMessage::Collation(
+					relay_parent,
+					para_id,
+					collator_id,
+				)) => {
+					self.handle_collation(relay_parent, para_id, collator_id)
+						.await;
+				}
+				ToJob::CandidateSelection(CandidateSelectionMessage::Invalid(
+					_,
+					candidate_receipt,
+				)) => {
+					self.handle_invalid(candidate_receipt).await;
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn handle_collation(
+		&mut self,
+		relay_parent: Hash,
+		para_id: ParaId,
+		collator_id: CollatorId,
+	) {
+		if self.seconded_candidate.is_none() {
+			let (candidate_receipt, pov) = match get_collation(
+				relay_parent,
+				para_id,
+				self.sender.clone(),
+			)
+			.await
+			{
+				Ok(response) => response,
+				Err(err) => {
+					log::warn!(target: "candidate_selection", "failed to get collation from collator protocol subsystem: {:?}", err);
+					return;
+				}
+			};
+
+			let pov = Arc::new(pov);
+
+			if !candidate_is_valid(
+				candidate_receipt.descriptor.clone(),
+				pov.clone(),
+				self.sender.clone(),
+			)
+			.await
+			{
+				return;
+			}
+
+			let pov = Arc::try_unwrap(pov)
+				.expect("only clone went to fn candidate_is_valid, which is complete; qed");
+
+			match second_candidate(
+				relay_parent,
+				candidate_receipt,
+				pov,
+				&mut self.sender,
+				&self.metrics,
+			)
+			.await
+			{
+				Err(err) => {
+					log::warn!(target: "candidate_selection", "failed to second a candidate: {:?}", err)
+				}
+				Ok(()) => self.seconded_candidate = Some(collator_id),
+			}
+		}
+	}
+
+	async fn handle_invalid(&mut self, candidate_receipt: CandidateReceipt) {
+		let received_from = match &self.seconded_candidate {
+			Some(peer) => peer,
+			None => {
+				log::warn!(target: "candidate_selection", "received invalidity notice for a candidate we don't remember seconding");
+				return;
+			}
+		};
+		log::info!(target: "candidate_selection", "received invalidity note for candidate {:?}", candidate_receipt);
```

Comment:
Nit: using a constant `TARGET` would make things a bit more compact and guarantee that spelling errors would not lead to log entries going to different targets.